### PR TITLE
Add factory to build RxActions

### LINKED ISSUE #3963: BindAction extension isn't working as you'd expect
## 🐛 Bug Report
In someplaces where the `BindAction` extension is being used, events are being forwarded to subscribers only once. This is causing the UI to become irresponsive after the first interaction with the user.

## Context
It seems to be affecting both android and iOS.

## Expected behavior
The events should be sent to the subscribers whenever they happen.

## To reproduce
Find a place where `BindAction` is being used, specially where the user ins't navigated to other screen after interaction.
Examples from where it was happening: 
Sign up screen -> Select Country; Sign up button
Settings screen -> Select default workspace

## Workaround & Related issues
Replacing the `BindAction` with a simple subscription does make it work.
Examples of the workaround here: #3952 #3953

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
This fixes an issue introduced when removing the async Task override of Subscribe, which caused RxAction components to be deallocated after the first completion, preventing the action to be executed twice.

Closes #3963

## Why do we want this?
We kind of want the app to work properly.

## How is it done?
As most weird and hard to solve problems this one has to do with threading. My guess is that await together with toObservable switches to some background thread resulting in some unexpected ordering of operations which causes the action components to be deallocated.

To solve this actions need to run on the main thread. To make it easier to create them without having to inject a scheduler all the time I created a `RxActionFactory` that's been injected to every VM that needs it.

### Why not another way?
Ideally we should find a way to access the main scheduler (and only that one) from every place in the app without having to inject anything all around, but that might not be easy due to parallel tests.

It's a bummer that we have to make our code more complicated because of tests, it should be like that. Any other ideas about this are welcome.

## :squid: Permissions
Be my guest

### CODE REVIEW

File: Toggl.Foundation.MvvmCross/ViewModels/Calendar/SelectUserCalendarsViewModel.cs [hunk 1]
```diff
@@ -15,18 +16,23 @@ namespace Toggl.Foundation.MvvmCross.ViewModels.Calendar
...
         public UIAction Done { get; private set; }
 
         public SelectUserCalendarsViewModel(
             IUserPreferences userPreferences,
             IInteractorFactory interactorFactory,
-            IMvxNavigationService navigationService)
-            : base(userPreferences, interactorFactory)
+            IMvxNavigationService navigationService,
+            IRxActionFactory rxActionFactory
+            )
```
Comment (daividssilverio) [inline Toggl.Foundation.MvvmCross/ViewModels/Calendar/SelectUserCalendarsViewModel.cs]: 👀

File: Toggl.Foundation.MvvmCross/ViewModels/BrowserViewModel.cs [hunk 2]
```diff
@@ -1,24 +1,33 @@
...
     {
         private readonly IMvxNavigationService navigationService;
+        private readonly IRxActionFactory rxActionFactory;
 
         public string Url { get; private set; }
 
         public string Title { get; private set; }
 
-        public BrowserViewModel(IMvxNavigationService navigationService)
+        public UIAction Close { get; }
```
Comment (daividssilverio) [inline Toggl.Foundation.MvvmCross/ViewModels/BrowserViewModel.cs]: Please fix the android usage for all the Close UIActions you are adding.
BrowserActivity.cs:
```
private void onNavigateBack(object sender, NavigationClickEventArgs e)
        {
            ViewModel.Close().Execute();
        }
```

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 3]
```diff
@@ -26,23 +29,27 @@ public sealed class SelectCountryViewModel : MvxViewModel<long?, long?>
         public IObservable<IEnumerable<SelectableCountryViewModel>> Countries { get; private set; }
 
         public InputAction<string> SetFilterText { get; }
+        public UIAction Close { get; }
```
Comment (daividssilverio) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: Same for this one, please fix the usage on`SelectCountryActivity.cs`.

File: Toggl.Foundation.MvvmCross/ViewModels/Settings/SettingsViewModel.cs [hunk 4]
```diff
@@ -92,6 +93,7 @@ public sealed class SettingsViewModel : MvxViewModel
         public UIAction PickDefaultWorkspace { get; }
         public UIAction SelectDurationFormat { get; }
         public UIAction SelectBeginningOfWeek { get; }
+        public UIAction Close { get; }
```
Comment (daividssilverio) [inline Toggl.Foundation.MvvmCross/ViewModels/Settings/SettingsViewModel.cs]: Same here, fix the `Close` usage on SettingsActivity.cs

**Comment by daividssilverio** [comment]
Another bug, the RxAction is getting disposed when a Dialog is cancelled by the user by clicking outside of bounds of the dialog (at least on Android). Example: try changing workspaces inside the Reports view then clicking outside of the dialog. iOS doesn't seem to be affected by it...
 
I'm fine with having this fixed in another issue if you say it's not related.

**Comment by Odrakir** [comment]
Superseded by #3995

File: Toggl.Foundation.MvvmCross/ViewModels/BrowserViewModel.cs [hunk 5]
```diff
@@ -1,24 +1,33 @@
...
+
+        public BrowserViewModel(IMvxNavigationService navigationService, IRxActionFactory rxActionFactory)
         {
             Ensure.Argument.IsNotNull(navigationService, nameof(navigationService));
+            Ensure.Argument.IsNotNull(rxActionFactory, nameof(rxActionFactory));
 
             this.navigationService = navigationService;
+            this.rxActionFactory = rxActionFactory;
```
[the remark this answers is not in the document]
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/ViewModels/BrowserViewModel.cs]: How about the `navigationService`? Does the closure capture it?

File: Toggl.Foundation.MvvmCross/FoundationExtensions.cs [hunk 6]
```diff
@@ -59,6 +60,7 @@ ITogglDataSource createDataSource(ITogglApi api)
 
                 Mvx.ConstructAndRegisterSingleton<IInteractorFactory, InteractorFactory>();
                 Mvx.ConstructAndRegisterSingleton<IAutocompleteProvider, AutocompleteProvider>();
+                Mvx.ConstructAndRegisterSingleton<IRxActionFactory, RxActionFactory>();
```
[the remark this answers is not in the document]
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/FoundationExtensions.cs]: I didn't think twice about this, just copied the way the interactor factory is created. Why should this be different?

[the remark this answers is not in the document]
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/FoundationExtensions.cs]: Oh, right, because of dependencies. I see now.

**Pull request closed without merging**